### PR TITLE
feat: angular handle route, component view/update permissions - FUI-2066

### PR DESCRIPTION
**Pull request by ArturKrasinski**
[Review Guidance](https://www.notion.so/genesisglobal/Platform-Code-Review-Process-ace93ba760cc4563b0dfb712e2a88d8a)

📓 &nbsp; **Related JIRA**

[FUI-2066](https://genesisglobal.atlassian.net/browse/FUI-2066)

🤔 &nbsp; **What does this PR do?**

- Add a not-permitted route (including a guard redirect if the user doesn't have permission to enter this route).
- Add an ```ngIf``` directive on components to check if the user has permission to view/edit data.
- Add an Angular component (error-message) and reuse it on the not-permitted route and within the component view.
- Lint and unify package.json with the version from web components / fix tests.

🚀 &nbsp; **Where should the reviewer start?**

- .genx/templates/angular
- client-tmp/angular

📑 &nbsp; **How should this be tested?**

### Simple Permission Example

```
rm -rf blankappseedtest && npx -y @genesislcap/genx@latest init blankappseedtest -x --ref ak/FUI-2066 --no-npm --framework Angular --apiHost "ws://public-foundation.genesislab.global/gwf/" --routes '[ { "name": "Home", "permissions": { "viewRight": null }, "tiles": [ { "title": "Grid", "type": "grid-pro", "config": { "resourceName": "ALL_TRADES", "permissions": { "viewRight": "viewTrades" }, "deferredGridOptions": { "columnDefs": [ { "field": "INSTRUMENT_NAME", "headerName": "Instrument Name" }, { "field": "VALUE", "headerName": "Inline coldef - VALUE" }, { "field": "QUANTITY", "headerName": "Quantity" }, { "field": "PNL", "headerName": "PNL" } ] } } }, { "title": "Form", "type": "smart-form", "config": { "resourceName": "EVENT_COUNTERPARTY_INSERT", "permissions": { "updateRight": "updateCounterparties" }, "uischema": { "type": "VerticalLayout", "elements": [ { "type": "Control", "label": "Inline form schema - main contact", "scope": "#/properties/MAIN_CONTACT" }, { "type": "Control", "label": "Issuer Name - Local Schema", "scope": "#/properties/ISSUER_NAME", "options": { "readonly": true } }, { "type": "Control", "label": "Price", "scope": "#/properties/PRICE" }, { "type": "Control", "scope": "#/properties/COUNTERPARTY", "options": { "allOptionsResourceName": "ALL_COUNTERPARTYS", "valueField": "COUNTERPARTY_ID", "labelField": "COUNTERPARTY_ID" } }, { "type": "Control", "label": "Password", "scope": "#/properties/PASSWORD", "options": { "isPassword": true } }, { "type": "Control", "label": "Password", "scope": "#/properties/PASSWORD", "options": { "textarea": true } } ] } } } ] }, { "name": "Realtime Dashboard", "permissions": { "viewRight": "viewDashboard" }, "tiles": [ { "title": "Entity manager tile", "type": "entity-manager", "config": { "resourceName": "ALL_COUNTERPARTYS", "permissions": { "viewRight": null, "updateRight": "updateDashboard" }, "title": "Counterparty Management", "updateEvent": "EVENT_COUNTERPARTY_MODIFY", "deleteEvent": "EVENT_COUNTERPARTY_DELETE", "createEvent": "EVENT_COUNTERPARTY_INSERT" } }, { "title": "Form tile", "type": "smart-form", "config": { "permissions": { "updateRight": null }, "resourceName": "EVENT_COUNTERPARTY_INSERT" } } ] }, { "name": "Analytics", "permissions": { "viewRight": "viewAnalytics" }, "tiles": [ { "title": "Charts Tile 2", "type": "chart", "config": { "permissions": { "viewRight": null }, "type": "pie", "resourceName": "ALL_POSITIONS", "xField": "INSTRUMENT_NAME", "yField": "VALUE" } }, { "title": "Charts Tile 3", "type": "chart", "config": { "permissions": { "viewRight": null }, "type": "column", "resourceName": "ALL_POSITIONS", "xField": "INSTRUMENT_ID", "yField": "VALUE" } } ] } ]' --csv '[{"name": "trade", "fields": ["a", "B"]}, {"name": "position", "fields": ["id", "TYPE"]} ]' --no-shell && cd blankappseedtest/client && npm run bootstrap && npm run dev
```

### More Complex Permission Example

```
rm -rf blankappseedtest && npx -y @genesislcap/genx@latest init blankappseedtest --ref ak/FUI-2066 --no-npm -x --framework Angular --routes '[
  {
    "name": "Home",
    "permissions": {
      "viewRight": null   
    },
    "tiles": [
      {
        "title": "Entity manager",
        "type": "entity-manager",
        "config": {
          "permissions": {
            "viewRight": "viewEM",    
            "updateRight": "updateEM"    
          },
          "modalPosition": "centre",
          "sizeColumnsToFit": true,
          "enableSearchBar": true,
          "resourceName": "ALL_POSITIONS",
          "title": "My Positions",
          "updateEvent": "EVENT_COUNTERPARTY_MODIFY",
          "deleteEvent": "EVENT_COUNTERPARTY_DELETE",
          "createEvent": "EVENT_COUNTERPARTY_INSERT",
          "createFormUiSchema": {
            "type": "VerticalLayout",
            "elements": [
              {
                "type": "Control",
                "label": "Inline create schema - main contact",
                "scope": "#/properties/MAIN_CONTACT"
              },
              {
                "type": "Control",
                "label": "Issuer Name - Local Schema",
                "scope": "#/properties/ISSUER_NAME",
                "options": {
                  "readonly": true
                }
              },
              {
                "type": "Control",
                "label": "Price",
                "scope": "#/properties/PRICE"
              },
              {
                "type": "Control",
                "scope": "#/properties/COUNTERPARTY",
                "options": {
                  "allOptionsResourceName": "ALL_COUNTERPARTYS",
                  "valueField": "COUNTERPARTY_ID",
                  "labelField": "COUNTERPARTY_ID"
                }
              },
              {
                "type": "Control",
                "label": "Password",
                "scope": "#/properties/PASSWORD",
                "options": {
                  "isPassword": true
                }
              },
              {
                "type": "Control",
                "label": "Password",
                "scope": "#/properties/PASSWORD",
                "options": {
                  "textarea": true
                }
              }
            ]
          },
          "updateFormUiSchema": {
            "type": "VerticalLayout",
            "elements": [
              {
                "type": "Control",
                "label": "Inline update schema - main contact",
                "scope": "#/properties/MAIN_CONTACT"
              },
              {
                "type": "Control",
                "label": "Issuer Name - Local Schema",
                "scope": "#/properties/ISSUER_NAME",
                "options": {
                  "readonly": true
                }
              },
              {
                "type": "Control",
                "label": "Price",
                "scope": "#/properties/PRICE"
              },
              {
                "type": "Control",
                "scope": "#/properties/COUNTERPARTY",
                "options": {
                  "allOptionsResourceName": "ALL_COUNTERPARTYS",
                  "valueField": "COUNTERPARTY_ID",
                  "labelField": "COUNTERPARTY_ID"
                }
              },
              {
                "type": "Control",
                "label": "Password",
                "scope": "#/properties/PASSWORD",
                "options": {
                  "isPassword": true
                }
              },
              {
                "type": "Control",
                "label": "Password",
                "scope": "#/properties/PASSWORD",
                "options": {
                  "textarea": true
                }
              }
            ]
          },
          "columns": [
            {
              "field": "INSTRUMENT_NAME",
              "headerName": "Instrument Name"
            },
            {
              "field": "VALUE",
              "headerName": "Inline coldef - VALUE"
            },
            {
              "field": "QUANTITY",
              "headerName": "Quantity"
            },
            {
              "field": "PNL",
              "headerName": "PNL"
            }
          ]
        }
      },
      {
        "title": "Grid",
        "type": "grid-pro",
        "config": {
          "resourceName": "ALL_TRADES",
          "permissions": {
            "viewRight": "viewTrades" 
          },
          "deferredGridOptions": {
            "columnDefs": [
              {
                "field": "INSTRUMENT_NAME",
                "headerName": "Instrument Name"
              },
              {
                "field": "VALUE",
                "headerName": "Inline coldef - VALUE"
              },
              {
                "field": "QUANTITY",
                "headerName": "Quantity"
              },
              {
                "field": "PNL",
                "headerName": "PNL"
              }
            ]
          }
        }
      },
      {
        "title": "Form",
        "type": "smart-form",
        "config": {
          "resourceName": "EVENT_COUNTERPARTY_INSERT",
          "permissions": {
            "updateRight": "updateCounterparties" 
          },
          "uischema": {
            "type": "VerticalLayout",
            "elements": [
              {
                "type": "Control",
                "label": "Inline form schema - main contact",
                "scope": "#/properties/MAIN_CONTACT"
              },
              {
                "type": "Control",
                "label": "Issuer Name - Local Schema",
                "scope": "#/properties/ISSUER_NAME",
                "options": {
                  "readonly": true
                }
              },
              {
                "type": "Control",
                "label": "Price",
                "scope": "#/properties/PRICE"
              },
              {
                "type": "Control",
                "scope": "#/properties/COUNTERPARTY",
                "options": {
                  "allOptionsResourceName": "ALL_COUNTERPARTYS",
                  "valueField": "COUNTERPARTY_ID",
                  "labelField": "COUNTERPARTY_ID"
                }
              },
              {
                "type": "Control",
                "label": "Password",
                "scope": "#/properties/PASSWORD",
                "options": {
                  "isPassword": true
                }
              },
              {
                "type": "Control",
                "label": "Password",
                "scope": "#/properties/PASSWORD",
                "options": {
                  "textarea": true
                }
              }
            ]
          }
        }
      }
    ]
  },
  {
    "name": "Realtime Dashboard",
    "permissions": {
      "viewRight": "viewDashboard" 
    },
    "tiles": [
      {
        "title": "Entity manager tile",
        "type": "entity-manager",
        "config": {
          "resourceName": "ALL_COUNTERPARTYS",
          "permissions": {
            "viewRight": null,
            "updateRight": "updateDashboard"
          },
          "title": "Counterparty Management",
          "updateEvent": "EVENT_COUNTERPARTY_MODIFY",
          "deleteEvent": "EVENT_COUNTERPARTY_DELETE",
          "createEvent": "EVENT_COUNTERPARTY_INSERT"
        }
      },
      {
        "title": "Form tile",
        "type": "smart-form",
        "config": {
          "permissions": {
            "updateRight": null
          },
          "resourceName": "EVENT_COUNTERPARTY_INSERT"
        }
      }
    ]
  },
  {
    "name": "Analytics",
    "permissions": {
      "viewRight": "viewAnalytics" 
    },
    "tiles": [
      {
        "title": "Grid Tile",
        "type": "grid-pro",
        "config": {
          "permissions": {
            "viewRight": null
          },
          "resourceName": "ALL_POSITIONS"
        }
      },
      {
        "title": "Charts Tile 1",
        "type": "chart",
        "config": {
          "permissions": {
            "viewRight": null
          },
          "type": "line",
          "resourceName": "ALL_POSITIONS",
          "xField": "INSTRUMENT_NAME",
          "yField": "VALUE"
        }
      },
      {
        "title": "Charts Tile 2",
        "type": "chart",
        "config": {
          "permissions": {
            "viewRight": null
          },
          "type": "pie",
          "resourceName": "ALL_POSITIONS",
          "xField": "INSTRUMENT_NAME",
          "yField": "VALUE"
        }
      },
      {
        "title": "Charts Tile 3",
        "type": "chart",
        "config": {
          "permissions": {
            "viewRight": null
          },
          "type": "column",
          "resourceName": "ALL_POSITIONS",
          "xField": "INSTRUMENT_ID",
          "yField": "VALUE"
        }
      }
    ]
  }
]' --csv '[{"name": "trade", "fields": ["a", "B"]}, {"name": "position", "fields": ["id", "TYPE"]} ]' --apiHost 'wss://public-foundation.genesislab.global/gwf/' --no-shell && cd blankappseedtest/client && npm run bootstrap && npm run dev
```

### Defaults

```
rm -rf blankappseedtest && npx -y @genesislcap/genx@latest init blankappseedtest -x --ref ak/FUI-2066 --no-npm
```

### Route and CSV parameter handling test

```
rm -rf blankappseedtest && npx -y @genesislcap/genx@latest init blankappseedtest --ref ak/FUI-2066 --no-npm -x --routes '[{"name":"Home","tiles":[{"title":"Entity manager","type":"entity-manager","config":{ "modalPosition": "centre", "sizeColumnsToFit": true, "enableSearchBar": true, "resourceName":"ALL_POSITIONS","title":"My Positions","updateEvent":"EVENT_COUNTERPARTY_MODIFY","deleteEvent":"EVENT_COUNTERPARTY_DELETE","createEvent":"EVENT_COUNTERPARTY_INSERT", "createFormUiSchema": {"type":"VerticalLayout","elements":[{"type":"Control","label":"Inline create schema - main contact","scope":"#/properties/MAIN_CONTACT"},{"type":"Control","label":"Issuer Name - Local Schema","scope":"#/properties/ISSUER_NAME","options":{"readonly":true}},{"type":"Control","label":"Price","scope":"#/properties/PRICE"},{"type":"Control","scope":"#/properties/COUNTERPARTY","options":{"allOptionsResourceName":"ALL_COUNTERPARTYS","valueField":"COUNTERPARTY_ID","labelField":"COUNTERPARTY_ID"}},{"type":"Control","label":"Password","scope":"#/properties/PASSWORD","options":{"isPassword":true}},{"type":"Control","label":"Password","scope":"#/properties/PASSWORD","options":{"textarea":true}}]}, "updateFormUiSchema": {"type":"VerticalLayout","elements":[{"type":"Control","label":"Inline update schema - main contact","scope":"#/properties/MAIN_CONTACT"},{"type":"Control","label":"Issuer Name - Local Schema","scope":"#/properties/ISSUER_NAME","options":{"readonly":true}},{"type":"Control","label":"Price","scope":"#/properties/PRICE"},{"type":"Control","scope":"#/properties/COUNTERPARTY","options":{"allOptionsResourceName":"ALL_COUNTERPARTYS","valueField":"COUNTERPARTY_ID","labelField":"COUNTERPARTY_ID"}},{"type":"Control","label":"Password","scope":"#/properties/PASSWORD","options":{"isPassword":true}},{"type":"Control","label":"Password","scope":"#/properties/PASSWORD","options":{"textarea":true}}]}, "columns": [{"field":"INSTRUMENT_NAME","headerName":"Instrument Name"},{"field":"VALUE","headerName":"Inline coldef - VALUE"},{"field":"QUANTITY","headerName":"Quantity"},{"field":"PNL","headerName":"PNL"}] }},{"title":"Grid","type":"grid-pro","config":{"resourceName":"ALL_TRADES", "deferredGridOptions": {"columnDefs":[{"field":"INSTRUMENT_NAME","headerName":"Instrument Name"},{"field":"VALUE","headerName":"Inline coldef - VALUE"},{"field":"QUANTITY","headerName":"Quantity"},{"field":"PNL","headerName":"PNL"}]} }},{"title":"Form","type":"smart-form","config":{"resourceName":"EVENT_COUNTERPARTY_INSERT", "uischema": {"type":"VerticalLayout","elements":[{"type":"Control","label":"Inline form schema - main contact","scope":"#/properties/MAIN_CONTACT"},{"type":"Control","label":"Issuer Name - Local Schema","scope":"#/properties/ISSUER_NAME","options":{"readonly":true}},{"type":"Control","label":"Price","scope":"#/properties/PRICE"},{"type":"Control","scope":"#/properties/COUNTERPARTY","options":{"allOptionsResourceName":"ALL_COUNTERPARTYS","valueField":"COUNTERPARTY_ID","labelField":"COUNTERPARTY_ID"}},{"type":"Control","label":"Password","scope":"#/properties/PASSWORD","options":{"isPassword":true}},{"type":"Control","label":"Password","scope":"#/properties/PASSWORD","options":{"textarea":true}}]} }}]}, {"name":"Realtime Dashboard","tiles":[{"title":"Entity manager tile","type":"entity-manager","config":{"resourceName":"ALL_COUNTERPARTYS","title":"Counterparty Management","updateEvent":"EVENT_COUNTERPARTY_MODIFY","deleteEvent":"EVENT_COUNTERPARTY_DELETE","createEvent":"EVENT_COUNTERPARTY_INSERT"}},{"title":"Form tile","type":"smart-form","config":{"resourceName":"EVENT_COUNTERPARTY_INSERT"}}]},{"name":"Analytics","tiles":[{"title":"Grid Tile","type":"grid-pro","config":{"resourceName":"ALL_POSITIONS"}},{"title":"Charts Tile 1","type":"chart","config":{"type":"line","resourceName":"ALL_POSITIONS","xField":"INSTRUMENT_NAME","yField":"VALUE"}},{"title":"Charts Tile 2","type":"chart","config":{"type":"pie","resourceName":"ALL_POSITIONS","xField":"INSTRUMENT_NAME","yField":"VALUE"}},{"title":"Charts Tile 3","type":"chart","config":{"type":"column","resourceName":"ALL_POSITIONS","xField":"INSTRUMENT_ID","yField":"VALUE"}}]}]' --csv '[{"name": "trade", "fields": ["a", "B"]}, {"name": "position", "fields": ["id", "TYPE"]} ]' --apiHost 'wss://public-foundation.genesislab.global/gwf/' --no-shell && cd blankappseedtest/client && npm run bootstrap && npm run dev
```

✅ &nbsp; **Demo**

#### route not permitted
image

#### component view/edit not permitted
image

#### npm run test
image

✅ &nbsp; **Checklist**

- [x] I have tested my changes.
- [x] I have added tests for my changes.
- [ ] I have updated the project documentation to reflect my changes.

### CODE REVIEW

File: .genx/templates/angular/entityManager.hbs [hunk 1]
```diff
@@ -1,45 +1,53 @@
...
-    {{#if config.sizeColumnsToFit}}
-    size-columns-to-fit
-    {{/if}}
-    {{#if config.enableSearchBar}}
-    enable-search-bar
-    {{/if}}
-></entity-management>
+<section style="contain: content; display: flex; height: 100%; width: 100%; overflow-y: auto; align-items: center; justify-content: center;">
+  <ng-container *ngIf="canUserViewElement('{{config.permissions.viewRight}}'); else notPermitted{{ config.index }}">
+    <entity-management
```
Comment (SzymonZur) [inline .genx/templates/angular/entityManager.hbs:3]: After running the application locally, entity-management is not visible - I haven't looked into it, but adding the following fixes it:

```
    width: 100%;
    height: 100%;
```

Video:

https://github.com/genesiscommunitysuccess/blank-app-seed/assets/22753085/3995b325-bd73-4a54-a5a6-a5a61bda9112


 note: I haven't tested grid-pro - I don't know if the same thing happens

**Review by SzymonZur — COMMENTED** [review]
Nice one! Permissions for components and routes work correctly

Screenshot 2024-07-10 at 21 54 18

Screenshot 2024-07-10 at 21 55 39

There is no permissions check for entity-management events - apart from that it looks great

File: .genx/templates/angular/entityManager.hbs [hunk 2]
```diff
@@ -1,45 +1,53 @@
...
+      design-system-prefix="rapid"
+      enable-row-flashing
+      enable-cell-flashing
+      {{#if config.title}}
+      title="{{ config.title }}"
+      {{/if}}
+      resourceName="{{ config.resourceName }}"
+      {{#if config.createEvent}}
+      createEvent="{{ config.createEvent }}"
```
[the remark this answers is not in the document]
Comment (ArturKrasinski) [reply .genx/templates/angular/entityManager.hbs]: fixed

File: .genx/templates/angular/entityManager.hbs (hunk 1 above)
Comment (ArturKrasinski) [reply .genx/templates/angular/entityManager.hbs:3]: css fixed

**Review by SzymonZur — APPROVED** [review]
Thanks for making changes, looks good now 

Screenshot 2024-07-12 at 10 07 48

Screenshot 2024-07-12 at 10 09 02

**Review by MrBrunoWolff — APPROVED** [review]

**Pull request merged**